### ISSUE TITLE
Adding .gensym. to a template does not make it local

### ISSUE THREAD

**Issue by winksaville** [nim-lang/Nim#2465]
```
template t() =
  template declX(str: string) {.gensym.} =
    var x {.inject.} : string = str

t()
declX("a string")
echo "x=", x
```

When compiling:

```
$ nim c -r --hints:off t
CC: t
[Linking]
/home/wink/prgs/nim/template/t
x=a string
```

**Comment by reactormonk** [comment]
From the manual:

> Whether a symbol that is declared in a template is exposed to the instantiation scope is controlled by the inject and gensym pragmas: gensym'ed symbols are not exposed but inject'ed are.

so it does exactly what you tell it to.

**Comment by goob** [comment]
@reactormonk 

I was wondering about this bug. I didn't think it was one at first, but I saw that Araq called it a bug in [this forum comment](http://forum.nim-lang.org/t/1100/2#6745).

Looking at it again, shouldn't the `gensym` pragma make the `declX` template local to the `t` template? Seems like it should produce: `undeclared identifier: declX`. That's what it does with a `proc` anyway.

```
template t() =
  proc declX(str: string) {.gensym.} =
    var x: string = str

t()
declX("a string")
echo "x=", x
```

**Comment by reactormonk** [comment]
Related: http://forum.nim-lang.org/t/1100

**Comment by winksaville** [comment]
In case you didn't realize, the reason you're getting:
```
test.nim(7, 11) Error: undeclared identifier: 'x'
```
Is because in your example you didn't have {.inject.} on the x:
```
var x {.inject.} : string = str
```
My full example is repeated here with the compile & run output:
```
$ cat t.nim; nim c -r --hints:off t.nim
template t() =
  template declX(str: string) {.gensym.} =
    var x {.inject.} : string = str

t()
declX("a string")
echo "x=", x
[Linking]
/home/wink/prgs/nim/template/t 
x=a string
```
As @goob said I would expect this to generate a compile error saying "declX" is undeclared. Instead it works.

**Comment by dom96** [comment]
The ``{.inject.}`` pragma has nothing to do with this bug. The following should also not work:

```nim
template t() =
  template foo() {.gensym.} =
    echo 42

t()
foo()
```

**Comment by timotheecour** [comment]
was about to file a new bug but I found this, looks like the same or similar:
```nim
import std/macros

macro mgen(): untyped =
  let name1 = genSym(nskTemplate, "mypriv1")
  let name2 = genSym(nskProc, "mypriv2")
  let name3 = genSym(nskLet, "mypriv3")
  result = newStmtList()
  result.add quote do:
    template `name1`() = echo "in template"
    proc `name2`() = echo "in proc"
    proc bobah() = echo "in bobah"
    let `name3` = 123
  echo repr(result)
  return result

mgen()

bobah() # ok, works
mypriv1()  # BUG: this shouldn't compile, it's genSym'd

when false:
  mypriv2() # inconsistent with mypriv1: this one doesn't compile
  echo mypriv3 # ditto
  mypriv1421054() # the gensym'd name doesn't compile either
  mypriv2421053() # ditto for mypriv2
```

**Comment by timotheecour** [comment]
the bug in top post is now fixed (as of 6aa971d39f1339a27a953972cd329f9cbe791f98 1.3.5 at least) but the bug in https://github.com/nim-lang/Nim/issues/2465#issuecomment-511076669 is still here

maybe a simple hacky fix would be to force gensym idents to start with `_`, although that still may not work eg if `mgen` is called twice, would those clash? unclear.

**Comment by metagn** [comment]
Now:

```nim
template t() =
  template foo() {.gensym.} =
    echo 42

t()
foo()
```

Gives:

```
(6, 1) Error: undeclared identifier: 'foo'
```
